### ISSUE TITLE
Out of disk space errors

### ISSUE THREAD

**Issue by brianmay** [Shippable/support#3681]
Hello,

We seem to be getting random out of disk space errors in CI, trying to create a ``SemLock``.

```
  File "/root/venv/3.6/lib/python3.6/site-packages/channels/test/liveserver.py", line 193, in _pre_setup
    server_ready = multiprocessing.Event()
  File "/usr/lib/python3.6/multiprocessing/context.py", line 92, in Event
    return Event(ctx=self.get_context())
  File "/usr/lib/python3.6/multiprocessing/synchronize.py", line 336, in __init__
    self._cond = ctx.Condition(ctx.Lock())
  File "/usr/lib/python3.6/multiprocessing/context.py", line 67, in Lock
    return Lock(ctx=self.get_context())
  File "/usr/lib/python3.6/multiprocessing/synchronize.py", line 163, in __init__
    SemLock.__init__(self, SEMAPHORE, 1, 1, ctx=ctx)
  File "/usr/lib/python3.6/multiprocessing/synchronize.py", line 60, in __init__
    unlink_now)
OSError: [Errno 28] No space left on device
```

For example:

https://app.shippable.com/bitbucket/wspdigital/myro-control-centre/runs/1599/summary/console
https://app.shippable.com/bitbucket/wspdigital/myro-control-centre/runs/1601/summary/console
https://app.shippable.com/bitbucket/wspdigital/myro-control-centre/runs/1605/summary/console

However, some builds are fine:

https://app.shippable.com/bitbucket/wspdigital/myro-control-centre/runs/1604/summary/console
https://app.shippable.com/bitbucket/wspdigital/myro-control-centre/runs/1606/summary/console
https://app.shippable.com/bitbucket/wspdigital/myro-control-centre/runs/1607/summary/console

Still trying to see if there is a pattern with the builds that work and don't work. However, it looks like maybe something is wrong on your end(?).

(Also looking at the above build logs I note that it isn't easy to tell at a glance - at least with the new interface - what triggered this build - e.g. pull request or commit? What branch?)

Thanks

**Comment by kanthchandra** [comment]
Hey @brianmay 

1. https://app.shippable.com/subs/bitbucket/wspdigital/nodes You can go to nodes page and click on the list of nodes available to see the usage statistics of the node. You can see something like this.
![selection_610](https://user-images.githubusercontent.com/1583444/27419482-d0927940-573e-11e7-90b9-ca6f933f3449.png)

> (Also looking at the above build logs I note that it isn't easy to tell at a glance - at least with the new interface - what triggered this build - e.g. pull request or commit? What branch?)
1. You can look at bar which has the runNumber at the top to see whether this build was because of a PR or commit or release.
 2. Also the branch name is available in the summary currently.
![selection_608](https://user-images.githubusercontent.com/1583444/27419501-e721b126-573e-11e7-9f77-2faee6452d01.png)

**Comment by ambarish2012** [comment]
@brianmay on which filesystem mount is Python creating the SemLock? Could you send us a simplified sample program for a repro that mirrors your test scenario? We can then clone that repository and run some tests.

Filesystem      Size  Used Avail Use% Mounted on
none            148G   86G   56G  61% /
tmpfs           1.9G     0  1.9G   0% /dev
tmpfs           1.9G     0  1.9G   0% /sys/fs/cgroup
/dev/xvda1      148G   86G   56G  61% /build
tmpfs           377M  448K  376M   1% /run
none            5.0M     0  5.0M   0% /run/lock
none            1.9G  244K  1.9G   1% /run/shm
none            100M     0  100M   0% /run/user
shm              64M     0   64M   0% /dev/shm

**Comment by brianmay** [comment]
Doing a google search suggests ``SemLock`` uses ``/dev/shm/`` by default.

I suspect this failure only occurs when we have 4 CI running at the same time, so might be hard to reproduce.

I find it odd that we would be using anything close to 64M, however will try to find out how much space we use from our end.

Thanks

**Comment by rageshkrishna** [comment]
If you're seeing this only when you have a large number of builds going on in quick succession, I think you might hitting the semaphore limit. I believe this error also results in an "out of space" exception.

Is it possible that your builds are not clearing out semaphores that they open? You could try adding a couple of commands to the `on_failure` section of your YML:

````yaml
build:
  on_failure:
    - ipcs -ls
    - ipcs -a
````

This will print the current semaphore limits and a summary of all open IPC structures, which will tell us if the limits are the problem.

**Comment by ambarish2012** [comment]
Very interesting @rageshkrishna. A question here - once the build completes and the build container is turned down, would the semlocks get released?

@brianmay in the free plan, the VM that is provisioned stays provisioned for an hour. So if you run a ton of builds, they run on the same machine, in different containers and possibly the leak might persist.

You could also try increasing the size of the shm mount - double or triple it. Seems like you have to script these steps - http://www.walkernews.net/2010/05/04/how-to-resize-devshm-filesystem-in-linux/.

**Comment by rageshkrishna** [comment]
@ambarish2012 Now that I think of it, you're right -- ideally, it _shouldn't_ affect the host and other containers. The `ipcs` output when it fails will be able to tell us for sure.

**Comment by brianmay** [comment]
Here is an example of a run with "ipcs" calls littered everywhere.

https://app.shippable.com/bitbucket/wspdigital/penguin_brian-myro-control-centre/runs/465/summary/console

I don't see anything obviously wrong here... Although it is possible all resources are freed after the tests quit. The tests don't appear to be leaking, however.

Will try again with even more debugging.

**Comment by brianmay** [comment]
It obviously failed here, but for some reason never ran the on failure stuff:

https://app.shippable.com/bitbucket/wspdigital/penguin_brian-myro-control-centre/runs/468/summary/console

Also noticed sometimes CI builds are much slower than normal, and doesn't seem to be any reason why. Adding the times gives a value significantly less than the total duration.

https://app.shippable.com/bitbucket/wspdigital/penguin_brian-myro-control-centre/runs/467/summary/console

**Comment by brianmay** [comment]
I don't see any mention of the branch name in the summary.
![screenshot 2017-06-26 10 13 13](https://user-images.githubusercontent.com/112729/27520892-3e343526-5a58-11e7-834f-71f0c8cf8820.png)
 From https://app.shippable.com/bitbucket/wspdigital/penguin_brian-myro-control-centre/runs/467/summary/console

**Comment by brianmay** [comment]
Tried again:

https://app.shippable.com/bitbucket/wspdigital/penguin_brian-myro-control-centre/runs/469/summary/console

For some reason ``on_failure`` is not executing. However, I don't see any evidence of any problems before that.

**Comment by brianmay** [comment]
If the tests don't fail with the out of disk space error, then the ``on_failure`` runs fine (exactly same git revision):

https://app.shippable.com/bitbucket/wspdigital/penguin_brian-myro-control-centre/runs/470/summary/console

So wonder if the out of disk space error is causing problems with the shippable process, too.

**Comment by Bharath92** [comment]
@brianmay Has there been a recent occurrence of this issue? If yes, then could you send us a simplified sample program in a repro that mirrors your test scenario? We can then clone that repository and run some tests.

**Comment by brianmay** [comment]
Sorry, completely forgot about this issue. Which is probably a good thing...

No, we haven't seen this problem again. The decision was made to disable our selenium tests (we haven't been able to get them to run reliably on shippable and debugging issues with tests isn't a priority at present time). So it is hard to be certain if the problem is now fixed or that we no longer trigger it because we disabled the selenium based tests (at around the same time).
